### PR TITLE
fix: disable multiplexing on macOS for some versions of curl

### LINKED ISSUE #12202: Cargo download dependency failed
### Problem

[Cargo download dependency failed](https://users.rust-lang.org/t/cargo-download-dependency-failed/94257)

This issue I filed on rust forum before, But I didn't get enough useful information.

### PR DESCRIPTION
**Pull request by weihanglo**
### What does this PR try to resolve?

Fixes #12202.
[Zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/curl.20error.20.60send.3A.20no.20filter.20connected.60).

In certain versions of libcurl when proxy is in use with HTTP/2
multiplexing, connections will continue stacking up. This was
fixed in libcurl 8.0.0 in curl/curl@821f6e2a89de8aec1c7da3c0f381b92b2b801efc

However, Cargo can still link against old system libcurl if it is from a
custom built one or on macOS. For those cases, multiplexing needs to be
disabled when those versions are detected.

### How should we test and review this PR?

The first commit refactors how Cargo gets `http.proxy` config, so that
wecan have the detection inside `Config`. Please call it out if you
feel uncomfortable to this.

The second one did the trick, and a unit test is added.

### Manual test the behavior

To manually test the behavior, you need to build cargo with
the `all-static` feature, which despite its name links libcurl
dynamically on macOS. It is a feature used by rustup distributions.

```
cargo b -F all-static
```

I don't really understand curl internals so had no luck to figure
out the relationship between proxies and HTTP/2 multiplexing.
I use [`hinata/nginx-forward-proxy`] to spin up a proxy to test curl's behavior.
It turns out that if there are fewer dependencies to download,
or we set `MAX_CONCURRENT_STREAMS` to a lower value in curl,
it largely affects the happening of the error `no filter connected`.

That implies multiplexing and http proxy are really influencing
each other in some way.

<details><summary>My patch to curl and curl-sys</summary>
<p>

```patch
diff --git a/Cargo.toml b/Cargo.toml
index ec9f878b3..3a687c309 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -205,3 +205,7 @@ vendored-libgit2 = ["libgit2-sys/vendored"]
 pretty-env-logger = ["pretty_env_logger"]
 # This is primarily used by rust-lang/rust distributing cargo the executable.
 all-static = ['vendored-openssl', 'curl/static-curl', 'curl/force-system-lib-on-osx']
+
+[patch.crates-io]
+curl = { git = "https://github.com/weihanglo/curl-rust", branch = "curlmopt" }
+curl-sys = { git = "https://github.com/weihanglo/curl-rust", branch = "curlmopt" }
diff --git a/src/cargo/sources/registry/http_remote.rs b/src/cargo/sources/registry/http_remote.rs
index 7e1f2a587..21ae3ca10 100644
--- a/src/cargo/sources/registry/http_remote.rs
+++ b/src/cargo/sources/registry/http_remote.rs
@@ -246,6 +246,8 @@ impl<'cfg> HttpRegistry<'cfg> {
                 .status("Updating", self.source_id.display_index())?;
         }
 
+        self.multi.set_max_concurrent_streams(10)?;
+
         Ok(())
     }
```

</p>
</details> 

[`hinata/nginx-forward-proxy`]: https://hub.docker.com/r/hinata/nginx-forward-proxy

### CODE REVIEW

**Comment by weihanglo** [comment]
BTW. Although this is not really a regression, with sparse registry the number of connections is easier to exceed `MAX_CONCURRENT_STREAMS` (default to 100 in libcurl). Should we backport it at least for <https://github.com/rust-lang/cargo/labels/beta-nominated>?

File: src/cargo/util/config/mod.rs [hunk 1]
```diff
@@ -1717,8 +1717,19 @@ impl Config {
...
-        self.http_config
-            .try_borrow_with(|| self.get::<CargoHttpConfig>("http"))
+        self.http_config.try_borrow_with(|| {
+            #[allow(unused_mut)]
+            let mut http = self.get::<CargoHttpConfig>("http")?;
+            #[cfg(target_os = "macos")]
+            {
+                // We only take care of macOS now, since it is one of the major
+                // platforms. For other OS we encourage not to stick to older
+                // versions of libcurl.
```
Comment (ehuss) [inline src/cargo/util/config/mod.rs:1727]: I'm a little uncertain about making this macos only. What would be the concern with just checking this on all platforms?

Comment (weihanglo) [reply src/cargo/util/config/mod.rs:1727]: That was just me trying to shave off all old system libcurl aggressively 😅.

The constraint has been removed.

And weird. I forced pushed something but it didn't show up in this pull request.

File: src/cargo/util/config/mod.rs [hunk 2]
```diff
@@ -2750,3 +2754,75 @@ impl Tool {
...
+) {
+    use crate::util::network;
+
+    if network::proxy::http_proxy_exists(http, config) && http.multiplexing.is_none() {
+        let bad_curl_versions = ["7.87.0", "7.88.0", "7.88.1"];
+        if bad_curl_versions
+            .iter()
+            .any(|v| curl_version.starts_with(v))
+        {
+            http.multiplexing = Some(false);
```
Comment (ehuss) [inline src/cargo/util/config/mod.rs:2780]: Can you add a log message here to indicate the downgrade? That might help with debugging issues. Perhaps something like:

```rust
log::info!("disabling multiplexing with proxy, curl version is {curl_version}");
```

`r=ehuss` with something like that added, thanks!

**Comment by ehuss** [comment]
If you want to backport this to beta, that sounds good to me!

I think at some point it would be good to set up a proxy test using the container tests.

**Comment by weihanglo** [comment]
Thanks for the review!

@bors r=ehuss

> I think at some point it would be good to set up a proxy test using the container tests.

Was thinking the same thing, though it feels a bit too hard to write one. I'll try when I have time. Thanks for the suggestion!

**Pull request merged**